ext/opentelemetry-instrumentation-starlette/src/opentelemetry/instrumentation/starlette/__init__.py

```diff
@@ -0,0 +1,82 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import Optional
+
+from starlette import applications
+from starlette.routing import Match
+
+from opentelemetry.ext.asgi import OpenTelemetryMiddleware
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.starlette.version import __version__  # noqa
+
+
+class StarletteInstrumentor(BaseInstrumentor):
+    """An instrumentor for starlette
+
+    See `BaseInstrumentor`
+    """
+
+    _original_starlette = None
+
+    @staticmethod
+    def instrument_app(app: applications.Starlette):
+        """Instrument a previously instrumented Starlette application.
+        """
+        if not getattr(app, "is_instrumented_by_opentelemetry", False):
+            app.add_middleware(
+                OpenTelemetryMiddleware,
```

Comment:
Just out of curiosity, are we creating a starlette instrumentation (even though we can use asgi) for the same reason we have a flask instrumentation (when we could have recommended wsgi)?

Reply:
Kind of. We need a language-specific instrumentation no matter what, to add the "http.route" span attribute (there is no generic way to express route either wsgi or asgi).

Flask has an extended reason to avoid calling updateName. However, I finally have an OTEP drafted that I believe will eliminate the deferred updateName requirement: https://github.com/open-telemetry/oteps/pull/115. At that point we can extend wsgi to create the flask instrumentation, in a very similar way to how this instrumentation works.